Handle name conflict with expected

gamesh411 opened the pull request:
In preparation for ASTImport Strategies, the interface of HandleNameConflict is changed to use Expected<DeclarationName>. FieldDecls and FunctionTemplateDecls are extended to use the name conflict resolution strategies.

Please double check the IDNS (identifier namespace) arguments that I have come up with, as I am not entirely familiar with the clang implementation of these (their semantic meaning is not clear to me).

Fixes #585 and #589

gamesh411 committed:
Make HandleNameConflict use Expected interface

gamesh411 committed:
Make ASTImport of FieldDecls respect HandleNameConflict

martong commented on lib/AST/ASTImporter.cpp:
```diff
@@ -5517,10 +5552,32 @@ ASTNodeImporter::VisitFunctionTemplateDecl(FunctionTemplateDecl *D) {
           }
           FoundByLookup = FoundTemplate;
           break;
+        } else {
```
I am not sure that we should have this `else` branch because function templates may overload. This means there will be structural in-equivalencies and that is okay here.
The `TODO` in this case is misleading.

martong replied:
However, we should check for structural eq in case of function template **sepcializations** (they do not overload).
See also https://github.com/Ericsson/clang/issues/585

gamesh411 replied:
Nice catch, thanks for the insight! Will update the PR to include the relevant comments.

gamesh411 committed:
Add comments to import function templates

The import of FunctionTemplates should not involve name conflict
resolution, as different function templates can overload.
FunctionTemplateSpecializations cannot overload, but the current
implementation of structural matching does not take the body of the
function into account, so there is no point checking.

martong reviewed (approved):
LG! Thanks, nice work!

martong commented:
https://reviews.llvm.org/D59692 includes this too.
